napari/_qt/qt_main_window.py

```diff
@@ -528,7 +528,10 @@ def add_dock_widget(
             # Keep the dropdown menus in the widget in sync with the layer model
             # if widget has a `reset_choices`, which is true for all magicgui
             # `CategoricalWidget`s
-            self.qt_viewer.viewer.layers.events.connect(widget.reset_choices)
+            layers_events = self.qt_viewer.viewer.layers.events
+            layers_events.inserted.connect(widget.reset_choices)
+            layers_events.removed.connect(widget.reset_choices)
+            layers_events.reordered.connect(widget.reset_choices)
```

Comment:
Probably for a different PR, but should we recurse into child widgets also?

Reply:
it already does that.  In the magicgui case `widget` is a `ContainerWidget`, and calling `reset_choices` on a container calls it on all of its children

Reply:
(more specifically, it is a `FunctionGui`, which is a subclass of `ContainerWidget`)

Reply:
>  In the magicgui case widget is a ContainerWidget, and calling reset_choices on a container calls it on all of its children

🤯

magicgui is indeed most deserving of its name. 😂